src/app/progress/quest.tsx

```diff
@@ -0,0 +1,72 @@
+import * as React from 'react';
+import * as _ from 'underscore';
+import { IDestinyObjectiveProgress, IDestinyInventoryComponent, IDestinyItemInstanceComponent, IDestinyItemComponent } from '../bungie-api/interfaces';
+import { sum } from '../util';
+import { t } from 'i18next';
+import { percent } from '../inventory/dimPercentWidth.directive';
+import './quest.scss';
+
+interface QuestProps {
+  defs;
+  item: IDestinyItemComponent;
+  objectives: IDestinyObjectiveProgress[];
+}
+
+export function Quest(props: QuestProps) {
+  const { defs, item, objectives } = props;
+
+  const itemDef = defs.InventoryItem.get(item.itemHash);
+
+  const percentComplete = sum(objectives, (objective) => {
+    const objectiveDef = defs.Objective.get(objective.objectiveHash);
+    if (objectiveDef.completionValue) {
+      return Math.min(1.0, (objective.progress || 0) / objectiveDef.completionValue) / objectives.length;
+    } else {
+      return 0;
+    }
+  });
+
+  return <div className="milestone-quest item-quest">
+    <div className="milestone-icon">
+      <img src={`https://www.bungie.net${itemDef.displayProperties.icon}`} />
+      {percentComplete > 0 &&
+        <span>{Math.floor(percentComplete * 100.0)}%</span>}
+    </div>
+    <div className="milestone-info">
+      <span className="milestone-name">{itemDef.displayProperties.name}</span>
+      <div className="milestone-description">{itemDef.displayProperties.description}</div>
+      <div className="quest-objectives">
+        {objectives.map((objective) =>
+          <Objective defs={defs} objective={objective} key={objective.objectiveHash}/>
+        )}
+      </div>
+    </div>
+  </div>;
+}
+
+interface ObjectiveProps {
+  defs;
+  objective: IDestinyObjectiveProgress;
+}
+function Objective(props: ObjectiveProps) {
+  const { defs, objective } = props;
+
+  const objectiveDef = defs.Objective.get(objective.objectiveHash);
+
+  const classNames = `objective-row ${objective.complete ? 'objective-complete' : ''} ${objectiveDef.completionValue === 1 ? 'objective-boolean' : ''}`
+
+  const displayName = objectiveDef.progressDescription ||
+      (objective.complete
+        ? t('Objectives.Complete')
+        : t('Objectives.Incomplete'));
+  const display = `${objective.progress || 0}/${objectiveDef.completionValue}`;
+
+  return <div className={classNames}>
+    <div className="objective-checkbox"><div></div></div>
```

Comment:
FYI you can just do `<div className="objective-checkbox"><div /></div>`

Reply:
That's just a bridge too far...  Weird that JSX would let you use bad HTML.

Reply:
Not bad HTML because JSX isn't HTML. `<div />` JSX still produces the valid  HTML `<div></div>`.

JSX strays in many places from 'pure HTML'  for various reasons, this being one of them. IMHO if you're going down the route of using JSX, you may as well use the syntactic sugars it provides.

Reply:
I'm fine with the attributes, but the markup that is in the JSX should be valid markup.  I get it that it isn't HTML 4, but it's using it, and we should follow its structural conventions since I have to read it.  Attribute camel casing isn't a bother as you can put whatever you want within the brackets normally.